File: DP(Dynamic_Programming, 동적계획법)/백준_11727_2xn타일링2/2xn타일링2.java
import java.util.*;

public class Main {
    static long[] arr = new long[1001];

    public static void main(String[] args) {
        arr[0] = 0; arr[1] = 1; arr[2] = 3;
        Scanner s = new Scanner(System.in);
        int n = s.nextInt();

        for(int i = 3; i<=n; i++){
            arr[i] = arr[i-1]  *2;
            if(i % 2 == 0) arr[i] += 1;
            else arr[i] -=1;
            arr[i]%= 10007;
        }
        System.out.println(arr[n]);

    }
}

File: DP(Dynamic_Programming, 동적계획법)/백준_11727_2xn타일링2/2xn타일링2.kt
import java.util.*
import kotlin.math.*

fun main(args: Array<String>){
    val arr = LongArray(1001)
    arr[0] = 0; arr[1] = 1; arr[2] = 3;
    val s = Scanner(System.`in`)
    val n = s.nextInt()

    for(i in 3..n){
        arr[i] = arr[i-1]*2
        if(i % 2 ==0 ) arr[i]++
        else arr[i]--
        arr[i] = arr[i] %10007
    }
    println(arr[n])

}